File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/findMode_501.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

import java.util.ArrayList;
import java.util.HashMap;

public class findMode_501 {
    HashMap<Integer, Integer> count = new HashMap<>();
    int max = 0;

    public int[] findMode(TreeNode root) {
        helper(root);
        ArrayList<Integer> l = new ArrayList<>();
        count.forEach((v, cnt) -> {
            if (cnt == max) l.add(v);
        });
        int[] res = new int[l.size()];
        for (int i = 0; i < l.size(); i++) {
            res[i] = l.get(i);
        }
        return res;
    }

    private void helper(TreeNode root) {
        if (root == null) return;
        int cnt = count.getOrDefault(root.val, 0) + 1;
        count.put(root.val, cnt);
        max = Math.max(cnt, max);
        helper(root.left);
        helper(root.right);
    }
}

File: Clion/ZeroTrac/Tree/findMode_501.cpp
//
// Created by TzashiNorpu on 10/31/2022.
//
#include "tree.h"

using namespace Tree;

vector<int> Solution::findMode(TreeNode *root) {
  unordered_map<int, int> freq_map;
  int max_freq = 0;
  vector<int> res;
  function<void(TreeNode *)> helper = [&](TreeNode *node) -> void {
      if (!node) return;
      freq_map[node->val]++;
      max_freq = max(max_freq, freq_map[node->val]);
      helper(node->left);
      helper(node->right);
  };
  for (auto &kv: freq_map)
    if (kv.second == max_freq) res.push_back(kv.first);
  return res;
}